### PR TITLE
fix: deserialize shorthand-tagged intrinsics in Fn::Not and Fn::Sub

### PR DESCRIPTION
**Pull request by darwinboersma**
Refs #694, aws/aws-cdk-cli#824

## Problem

Migrating CloudFormation templates that use YAML shorthand intrinsics in two common places fails with:

```
... could not be generated because <path>: untagged and internally tagged enums do not support enum input
```

Two spots in the parser model their payloads as `#[serde(untagged)]` enums:

- `Singleton` — the operand of `Fn::Not`
- `SubPayload` — the `Fn::Sub` `[template, {variables}]` form

serde's untagged deserialization buffers the input through `Content` and then refuses *enum* input, but `serde_yaml` surfaces YAML shorthand tags (`!Equals`, `!Ref`, `!GetAtt`, …) as enum input. So a shorthand intrinsic nested under `Fn::Not`, or inside a `Fn::Sub` variables map, fails to parse — e.g.:

```yaml
# condition operand
!Not [!Equals [!Ref Foo, ""]]
# Sub variables map
!Sub ["${Foo}", { Foo: !Ref Bar }]
```

## Fix

Replace both untagged enums with hand-written `Deserialize` impls that deserialize the inner `ConditionValue` / `ResourceValue` directly. Those types already have visitors that understand YAML tags, so deserializing them directly avoids the `Content` buffering that rejected the tagged input. Behavior for the previously-working forms (bare value, single-element list, `[template, {literals}]`) is preserved.

This is parser-only and language-agnostic — no synthesizer, IR, or schema changes.

## Scope

This fixes the *deserialization* failure. It does not change which intrinsic functions are permitted *inside* a `Conditions` block — e.g. `Fn::Join`/`Fn::Sub` used inside a condition is a separate gap. That's why this references #694 rather than closing it.

## Testing

- Added regression tests `function_not_nested_shorthand` (conditions) and `intrinsic_sub_shorthand_tagged_variable` (resources).
- `cargo test --all-features` — all tests pass.
- `cargo fmt --check` and `cargo clippy --tests -- -Adead-code -Dwarnings -Dclippy::dbg_macro` — clean.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/parser/condition/mod.rs [hunk 1]
```diff
@@ -249,19 +249,84 @@ impl<'de> serde::Deserialize<'de> for ConditionValue {
...
+            }
+            fn visit_i64<E: Error>(self, val: i64) -> Result<Self::Value, E> {
+                Ok(ConditionValue::String(val.to_string()))
+            }
+            fn visit_u64<E: Error>(self, val: u64) -> Result<Self::Value, E> {
+                Ok(ConditionValue::String(val.to_string()))
+            }
+            fn visit_f64<E: Error>(self, val: f64) -> Result<Self::Value, E> {
+                Ok(ConditionValue::String(val.to_string()))
+            }
```
Comment (mrgrain) [inline src/parser/condition/mod.rs]: These arms are a second copy of how `ConditionValue` already handles scalars, so the two can quietly drift apart over time. Can we delegate to `ConditionValue` here the way the `visit_map`/`visit_enum` arms just below do, instead of rebuilding the strings by hand? If there's no clean way to delegate the owned-scalar cases, let's at least add tests that hit each of these arms (bare bool/int/float/string operands to `!Not`) so we'd catch a divergence.

File: src/parser/condition/mod.rs [hunk 2]
```diff
@@ -249,19 +249,84 @@ impl<'de> serde::Deserialize<'de> for ConditionValue {
...
+            fn visit_seq<A: serde::de::SeqAccess<'de>>(
+                self,
+                mut seq: A,
+            ) -> Result<Self::Value, A::Error> {
+                let value: ConditionValue = match seq.next_element()? {
+                    Some(value) => value,
+                    None => return Err(A::Error::invalid_length(0, &self)),
+                };
+                // `Fn::Not` takes exactly one operand; drain any extras.
+                while seq.next_element::<serde::de::IgnoredAny>()?.is_some() {}
```
Comment (mrgrain) [inline src/parser/condition/mod.rs]: Let's not silently swallow extra operands here. `Fn::Not` takes exactly one, so `!Not [a, b]` is a broken template and quietly dropping `b` hides the mistake from the user. Please return an error when there's a trailing element instead of draining it.

File: src/parser/intrinsics/mod.rs [hunk 3]
```diff
@@ -214,20 +214,52 @@ impl StringOrPair {
...
+                self,
+                mut seq: A,
+            ) -> Result<Self::Value, A::Error> {
+                let template: String = match seq.next_element()? {
+                    Some(template) => template,
+                    None => return Err(A::Error::invalid_length(0, &self)),
+                };
+                // The second element (the variables map) is optional.
+                let variables: Option<ResourceValue> = seq.next_element()?;
+                while seq.next_element::<serde::de::IgnoredAny>()?.is_some() {}
```
Comment (mrgrain) [inline src/parser/intrinsics/mod.rs]: Same as the `Fn::Not` case: a `Fn::Sub` with 3+ elements is malformed, and draining the extras hides that. Let's reject the trailing elements with an error rather than ignoring them.

**Review by mrgrain — COMMENTED** [review]
A couple of inline suggestions on the new visitors.

**Review by mrgrain — CHANGES_REQUESTED** [review]
Nice and thank you! Please increase test coverage for the changed files to meet threshold and resolve my requests. This should be mergeable soon.

File: src/parser/condition/mod.rs (hunk 1 above)
Comment (darwinboersma) [reply src/parser/condition/mod.rs]: Good call, sharing that handling of scalars seems ideal and saves some boilerplate! I'll dig into that now

**Comment by darwinboersma** [comment]
Thanks for the review! Addressed all three in acf7cba:

1. **Scalar arms duplicating `ConditionValue`** — the non-list `Fn::Not` operand now delegates to `ConditionValue` for every form (scalars route through `serde::de::value` deserializers, the same way the `visit_map`/`visit_enum` arms already did), so there's a single source of truth and no risk of drift.
2. **`Fn::Not` trailing operands** — now returns a deserialization error instead of draining them; `!Not [a, b]` (and `!Not []`) are rejected.
3. **`Fn::Sub` trailing elements** — same treatment; a 3rd element is now an error rather than silently dropped.

Also added tests covering the delegated scalar/map/enum arms and the new empty-list / extra-element error paths, which should bring patch coverage back over the threshold.

**Review by mrgrain — APPROVED** [review]
